torch/csrc/jit/fuser/compiler.cpp

```diff
@@ -144,6 +156,111 @@ static void setInputBroadcastGroups(KernelSpec& spec) {
       std::back_inserter(spec.inputBroadcastGroups()));
 }
 
+// This function moves _grad_sum_to_size nodes along the computation graph
+// of the fusion group to the outputs and then records the shape inputs
+// in order for summation to be applied after the kernel.
+// Note that the correctness relies on the invariant that
+// _grad_sum_to_size is only applied to gradient nodes created by autodiff.
+// This is important because it ensures that in the mul and div nodes only
+// one argument (in the case of diff the numerator) has a summed value.
+// If two arguments to mul had one, we would be in trouble, but thanks
+// to the chain rule, we're OK.
+void processGradSumToSize(KernelSpec& spec) {
+  auto graph = spec.graph();
+
+  // keep track of one sumToSize hitting each output
+  auto& outputMapAndSizes = spec.outputMapAndSizes();
+  JIT_ASSERT(outputMapAndSizes.empty());
+  std::vector<int64_t> outputGradSumToSizes(graph->outputs().size(), -1);
+
+  static OperatorSet commutes_with_SumToSize{{
+      "aten::mul(Tensor self, Tensor other) -> Tensor",
+      "aten::div(Tensor self, Tensor other) -> Tensor",
+      // for div we might check whether we're the first argument
+      "aten::mul(Tensor self, Scalar other) -> Tensor",
+      "aten::div(Tensor self, Scalar other) -> Tensor",
+      "aten::neg(Tensor self) -> Tensor",
+      "aten::add(Tensor self, Tensor other, *, Scalar alpha) -> Tensor",
+      // add this used to be prim::AutogradAdd
+  }};
+  // Scan the graph. We will delete nodes. We want later (in the graph)
+  // _grad_sum_to_size nodes to have priority over earlier ones. Thus
+  // we scan backwards.
+  for (auto it = graph->nodes().rbegin(); it != graph->nodes().rend(); it++) {
+    auto* node = *it;
+    if (node->kind() == aten::_grad_sum_to_size) {
+      use_list uses_to_process(node->output()->uses());
+      while (!uses_to_process.empty()) {
+        auto user = uses_to_process[0].user;
+        auto offset = uses_to_process[0].offset;
+        uses_to_process.erase(uses_to_process.begin());
+        if (user->matches(
+                "aten::type_as(Tensor self, Tensor other) -> Tensor")) {
+          // sometimes, a mask or similar is cast to the same
+          // type as the gradient. But we only want to allow that as offset==1
+          // (i.e. the type)
```

Comment:
Sorry, why do we only want to allow the output of `_grad_sum_to_size` to be used as a template for a cast and not as `self`? That seems like a too strong condition for an assert.

Reply:
Indeed, I'll absorb sumtosize hitting other and pass through self.